### PR TITLE
Added Metadata definition with Text and Option

### PR DESCRIPTION
**Pull request by jemrayfieldft**
Metadata definition.
I will add create project (group) in a separate branch

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Mio Config
-
+.
```
Comment (jspc) [inline README.md]: Whats this?

File: Gemfile [hunk 2]
```diff
@@ -13,3 +13,4 @@ gem 'faraday'
 gem 'faraday-detailed_logger'
 gem 'net-http-persistent'
 gem 'rake'
+gem 'builder'
```
Comment (jspc) [inline Gemfile]: Could you put this alphabetical?

File: lib/mio/model.rb [hunk 3]
```diff
@@ -69,6 +69,7 @@ def configure
                         config_hash
     end
 
+
```
Comment (jspc) [inline lib/mio/model.rb]: Extra line break?

File: lib/mio/client.rb [hunk 4]
```diff
@@ -87,8 +96,8 @@ def post url, payload, opts
       Mio::Requests.make_request :post, @agent, url, opts, payload
     end
 
-    def put url, payload, opts
-      Mio::Requests.make_request :put, @agent, url, opts, payload
+    def put url, payload, opts, content_type=nil
```
Comment (jspc) [inline lib/mio/client.rb]: Perhaps default this to `+json`

File: lib/mio/model/metadataDefinition/definition.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+class Mio
+  class Model
+    class MetadataDefinition
+      class Definition < Model
```
Comment (jspc) [inline lib/mio/model/metadataDefinition/definition.rb]: So this is a MetadataDefinition Definition?

File: lib/mio/model/metadataDefinition/definition.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        field :formType, String, 'Form control type'
+        field :validationHandler, String, 'Validation handler', nil
+        field :options, Array, 'Array of options', nil
+
+        nested true
+
+      end
+    end
+  end
+end
```
Comment (jspc) [inline lib/mio/model/metadataDefinition/definition.rb]: Missing linebreak

File: lib/mio/model/metadatadefinition/option.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        field :displayName, String, 'Metadata Definition Description'
+        field :default, Symbol, 'Metadata type'
+        field :value, String, 'Indexed and searchable', :true
+
+        nested true
+
+      end
+    end
+  end
+end
```
Comment (jspc) [inline lib/mio/model/metadatadefinition/option.rb]: Linebreak again

File: spec/lib/mio/model/metadatadefinition_spec.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  context 'when instantiated with valid options' do
+    let(:metadata_definition){subject.new(client, model_args)}
+    [:definition_xml, :trim_xml_from_element].each do |m|
+      it "should respond to ##{m.to_s}" do
+        expect(metadata_definition).to respond_to(m)
+      end
+    end
+  end
+
+end
```
Comment (jspc) [inline spec/lib/mio/model/metadatadefinition_spec.rb]: Linebreak police 👮  whew whew

File: lib/mio/model/metadataDefinition/definition.rb (hunk 6 above)
Comment (jemrayfieldft) [inline lib/mio/model/metadataDefinition/definition.rb]: You got a standard doc you are following.

Where do you expect the line break here... 
Like this:

class Mio
  class Model
    class MetadataDefinition
      class Definition < Model
        set_resource :metadatadefinition

        field :name, String, 'Metadata Definition Name'
        field :description, String, 'Metadata Definition Description'
        field :type, String, 'Metadata type'
        field :searchable, Symbol, 'Indexed and searchable', :true
        field :editable, Symbol, 'Editable field', :true
        field :required, Symbol, 'Required mandatory metadata item', :true
        field :formType, String, 'Form control type'
        field :validationHandler, String, 'Validation handler', nil
        field :options, Array, 'Array of options', nil

        nested true

      end

    end
  end
end

Let us know

File: lib/mio/model/metadatadefinition/option.rb (hunk 7 above)
Comment (jemrayfieldft) [inline lib/mio/model/metadatadefinition/option.rb]: As above... not sure where you want the linebreak

File: spec/lib/mio/model/metadatadefinition_spec.rb (hunk 8 above)
Comment (jemrayfieldft) [inline spec/lib/mio/model/metadatadefinition_spec.rb]: Err... mr policeman can you tell me wot I have done wrong....

please mr policeman.

File: lib/mio/model/metadataDefinition/definition.rb (hunk 6 above)
Comment (jspc) [inline lib/mio/model/metadataDefinition/definition.rb]: If you see the diff you'll see a warning icon telling you theres a missing linebreak at the end of the file

File: lib/mio/model/metadatadefinition/option.rb (hunk 7 above)
Comment (jspc) [inline lib/mio/model/metadatadefinition/option.rb]: End of file, see the diff

File: lib/mio/model/metadataDefinition/definition.rb (hunk 5 above)
Comment (jemrayfieldft) [inline lib/mio/model/metadataDefinition/definition.rb]: Aye thats how it is in mio model

File: spec/lib/mio/model/metadatadefinition_spec.rb (hunk 8 above)
Comment (jspc) [inline spec/lib/mio/model/metadatadefinition_spec.rb]: As above, above - see diff

Comment (jemrayfieldft) [inline spec/lib/mio/model/metadatadefinition_spec.rb]: Good to merge mr line break policeman?

Comment (jspc) [inline spec/lib/mio/model/metadatadefinition_spec.rb]: No- there is still a missing EOL

File: Gemfile [hunk 9]
```diff
@@ -8,8 +8,10 @@ group :development do
   gem 'vcr'
 end
 
+gem 'builder'
 gem 'colorize'
 gem 'faraday'
 gem 'faraday-detailed_logger'
 gem 'net-http-persistent'
 gem 'rake'
+
```
Comment (jspc) [inline Gemfile]: I think you've been a little overzealous adding breaks

File: lib/mio/model/metadataDefinition/definition.rb (hunk 5 above)
Comment (jspc) [inline lib/mio/model/metadataDefinition/definition.rb]: I don't like that- it makes no sense, but I don't care enough to block it

File: lib/mio/model/metadata_definition.rb [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        if (endTagIdx)
+          xmlEnd = endTagIdx + endTag.length
+          if (xmlEnd < xml.length)
+            xml = xml[0...-(xml.length-xmlEnd)]
+          end
+        end
+        xml
+      end
+
+      def definition_xml
```
Comment (jspc) [inline lib/mio/model/metadata_definition.rb]: Why not `definition_hash` like the others?

File: lib/mio/model/metadata_definition.rb [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                     end
+                   end
+               end
+             end
+          end
+        end
+        trim_xml_from_element builder.target!
+      end
+
+      def go
```
Comment (jspc) [inline lib/mio/model/metadata_definition.rb]: This is exactly the same as the others, no? Just with xml?

File: lib/mio/model/metadata_definition.rb [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+          raise Mio::Model::EmptyField, 'Field definitions to Mio::Model::MetadataDefinition must contain at least one definition'
+        end
+
+        unless look_up
+          @object = create
+        else
+          @object = look_up
+          set_start :stop
+        end
+        @definition_path = "#{self.class.resource_name}/#{@object['id']}/definition"
```
Comment (jspc) [inline lib/mio/model/metadata_definition.rb]: Does this need to be an instance variable? I can't see where it is used (apart from as a reference) in other scopes

File: lib/mio/model/metadatadefinition/option.rb [hunk 13]
```diff
@@ -0,0 +1,17 @@
+class Mio
+  class Model
+    class MetadataDefinition
+      class Option < Model
+        set_resource :metadatadefinition
+
+        field :name, String, 'Metadata Definition Name'
+        field :displayName, String, 'Metadata Definition Description'
+        field :default, Symbol, 'Metadata type'
+        field :value, String, 'Indexed and searchable', :true
```
Comment (jspc) [inline lib/mio/model/metadatadefinition/option.rb]: This is a string but defaults to a symbol- something isn't right there

File: lib/mio/model/metadata_definition.rb (hunk 10 above)
Comment (jemrayfieldft) [inline lib/mio/model/metadata_definition.rb]: Because it creates xml not a hash

File: lib/mio/model/metadata_definition.rb (hunk 11 above)
Comment (jemrayfieldft) [inline lib/mio/model/metadata_definition.rb]: no it uses client.definition which requires xml and a different Content-Type

File: lib/mio/model/metadatadefinition/option.rb (hunk 13 above)
Comment (jemrayfieldft) [inline lib/mio/model/metadatadefinition/option.rb]: default is crap

File: lib/mio/model/metadata_definition.rb (hunk 10 above)
Comment (jspc) [inline lib/mio/model/metadata_definition.rb]: If you create 'definition_type :xml' in this file somewhere (like we do with stuff like with `nested true` or `set_resource :blah`) then you make this a lot more portable, and a ;lot simpler to swap out. Then your config hash stays as an object and transforming it to xml happens in the request in the same way transforming to json does.

Comment (jemrayfieldft) [inline lib/mio/model/metadata_definition.rb]: The xml transformation is specific to metadata definition, why am I moving it out and attempting to make it generic "happens in request".

**Comment by jemrayfieldft** [comment]
Should be re-factored when mio 5.3.0 is availible to use json Content-Type

**Pull request merged**